### ISSUE TITLE
Privacy Pass

### ISSUE THREAD

**Issue by oczkers** [VeNoMouS/cloudscraper#160]
It would be much better to resolve one captcha and use passes when needed instead of 30 captchas that takes a lot of time. Did You looked into [Privacy Pass](https://chrome.google.com/webstore/detail/privacy-pass/ajhmfdgkijocedmfjonnpjfojldioehi) algorithm?

> every time the Privacy Pass plugin needs a new set of privacy passes, it creates a set of thirty random numbers t1 to t30, hashes them into an elliptic curve (P-256 in our case), blinds them with a value b and sends them along with a challenge solution. The server returns the set of points multiplied by its private key and a batch discrete logarithm equivalence proof. Each pair (ti, HMACi(M)) constitutes a Privacy Pass and can be redeemed to solve a subsequent challenge. Voila!

**Comment by Eastkap** [comment]
Privacy pass is already disabled on some sites and would probably be completely abandoned if implemented here imo

**Comment by VeNoMouS** [comment]
I was reading up on it, it has to implemented server side from what i was see'ing it isn't instantly deployed by Cloudflare.... I mean I could be wrong on this??!?
![source](https://user-images.githubusercontent.com/597545/76806278-cd065d80-6845-11ea-8205-20a50447852b.gif)

**Comment by oczkers** [comment]
Yes, we need private key used to generate it fully automatic and i assume they guard it better than cookies ;-).

But probably if we manage to interact with privacy pass extension it could be at least "semi-automatic" - my logic is this:
1. launch/enable privacy pass extension (using whole chrome/firefox browser probably)
2. init privacy passes generation
3. solve recaptcha - privacy pass generates 30 hashes here
4. retrieve all 30 hashes probably by faking connection between privacy pass and cloudflare site so they don't mark it as used
5. use privacy pass hashes for whatever site we want

Of course this is completely different solution than cloudscraper so it's not exactly enhancement and i havn't checked if this is even possible, just looking around :-)

**Comment by VeNoMouS** [comment]
I have no problem starting this as a project, perhaps as a completely separate repo, and just integrating in to **cloudscraper** once I have a working solution

**Comment by Eric898989** [comment]
Definitely possible :-)
You just modify the chrome extension to send the tokens received from the website to a local sql server (what I did, and the extension is also using the sql database instead of the local storage of the browser), then from your app you can simply consume these tokens in generating redeem headers.
However, it seems that some cloudflare protected websites have put counter measures to prevent using it "freely", even if you got a cf_clearance cookie from bypass token, the next page you request you get again a hcaptcha, if you pass it then you are cleared!!
